hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java

```diff
@@ -6912,69 +6886,58 @@ else if ( bindValueCount < bindValuePaddingCount ) {
 		String separator = NO_SEPARATOR;
 
 		final Iterator<Expression> iterator = listExpressions.iterator();
-		int itemNumber = 0;
-		while ( iterator.hasNext() && ( inExprLimit == 0 || itemNumber < inExprLimit ) ) {
-			final Expression listExpression = itemAccessor.apply( iterator.next() );
+		Expression listExpression = null;
+		for ( int i = 0; i < bindValueCountWithPadding; i++ ) {
+			if ( inExprLimit > 0 && i % inExprLimit == 0 && i != 0 ) {
+				appendInClauseSeperator( inListPredicate );
+				separator = NO_SEPARATOR;
+			}
+
+			if ( iterator.hasNext() ) { // If the iterator is exhausted, reuse the last expression for padding.
+				listExpression = itemAccessor.apply( iterator.next() );
+			}
+			// The only way for expression to be null is if listExpressions is empty,
+			// but if that is the case the code takes an early exit.
+			assert listExpression != null;
+
 			appendSql( separator );
 			listExpression.accept( this );
 			separator = COMMA_SEPARATOR;
-			itemNumber++;
-			// If we encounter an expression that is not a parameter or literal, we reset the inExprLimit and bindValueMaxCount
-			// and just render through the in list expressions as they are without padding/splitting
+
+			// If we encounter an expression that is not a parameter or literal, we reset the inExprLimit and
+			// bindValueMaxCount and just render through the in list expressions as they are without padding/splitting
 			if ( !( listExpression instanceof JdbcParameter || listExpression instanceof SqmParameterInterpretation || listExpression instanceof Literal ) ) {
 				inExprLimit = 0;
-				bindValueMaxCount = bindValueCount;
+				bindValueCountWithPadding = bindValueCount;
 			}
 		}
 
-		if ( itemNumber != inExprLimit && bindValueCount == bindValueMaxCount ) {
+		appendSql( CLOSE_PARENTHESIS );
+		if ( parenthesis ) {
 			appendSql( CLOSE_PARENTHESIS );
-			return;
 		}
+	}
 
-		if ( inExprLimit > 0 && bindValueCount > inExprLimit ) {
-			do {
-				if ( inListPredicate.isNegated() ) {
-					append( ") and " );
-					inListPredicate.getTestExpression().accept( this );
-					appendSql( " not" );
-				}
-				else {
-					append( ") or " );
-					inListPredicate.getTestExpression().accept( this );
-				}
-				appendSql( " in (" );
-				separator = NO_SEPARATOR;
-				itemNumber = 0;
-				while ( iterator.hasNext() && itemNumber < inExprLimit ) {
-					final Expression listExpression = iterator.next();
-					appendSql( separator );
-					itemAccessor.apply( listExpression ).accept( this );
-					separator = COMMA_SEPARATOR;
-					itemNumber++;
-				}
-			} while ( iterator.hasNext() );
+	private void appendInClauseSeperator(InListPredicate inListPredicate) {
```

Comment:
```suggestion
	private void appendInClauseSeparator(InListPredicate inListPredicate) {
```

Reply:
I squashed the fix for this typo into the first commit.